### PR TITLE
Show a progress indicator when opening a PDE run configuration.

### PR DESCRIPTION
**Pull request by fedejeanne**
The progress indicator uses `IProgressMonitor.UNKNOWN` as the total amount of work because the (sometimes) long running operation is deep in the call stack: it goes all the way down to the `TracingOptionsManager` in `org.eclipse.pde.core` (dependency issues).

The operation can **not** be canceled.

Fixes https://github.com/eclipse-pde/eclipse.pde/issues/679

### CODE REVIEW

**Comment by HannesWell** [comment]
With this change now always the progress-indicator pops when opening the launch-configuration Editor, even if it is just for a very short time.
A pop-up that vanishes within the fraction of a second so that the user cannot read its content is IMHO disturbing.
Wouldn't it be possible to show the pop-up only after a certain amount of time or estimate better if the operation is long-runnning?

**Comment by fedejeanne** [comment]
> Wouldn't it be possible to show the pop-up only after a certain amount of time or estimate better if the operation is long-runnning?

I made some changes and shifted all the heavy lifting to the **activation** of the _Tracing_ tab, so I managed to get rid of the dialog at the beginning. 

The problem with the flickering still remains, but it now happens while changing some of the configurations in the _Tracing_ tab and the flickering happens in the lower part of the _Run configurations..._ dialog (in the progress bar). The good thing is that fixing that is easier than fixing the dialog, here's a draft PR to do that --> https://github.com/eclipse-platform/eclipse.platform/pull/591

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java
Comment (HannesWell) [inline ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java:69]: I don't think maintaining an own cache is not necessary since the TracingOptionsManager already caches the template internally and since `TracingOptionsManager.getTracingTemplate()` seem to be the expensive method (it reads the tracing options from all plugins, which is relatively heavy IO), a second cache seem unnecessary.

**Review by HannesWell — COMMENTED** [review]
I looked into this and as written below an additional cache seems not to be necessary.

But shifting the first call to the TracingManager seems reasonable.
I suggest we add a method to `TracingOptionsManager` like `ensureInitialized()` that basically calls getTracingTemplate() and passes a `ProgressMonitor` to it so that we can at least see progress. It shouldn't be canclable because otherwise the state would become inconsistent or more complex to manage.

I'm currently working on a greater clean up of many of the launcher classes, which would also make the TracingOptionsManager a bit simpler. Maybe you want to await that before adding the suggested `ensureInitialized()` method, in order to avoid conflicts.

**Comment by fedejeanne** [comment]
> I'm currently working on a greater clean up of many of the launcher classes, which would also make the TracingOptionsManager a bit simpler. Maybe you want to await that before adding the suggested ensureInitialized() method, in order to avoid conflicts.

👌 
Please let me know when your PR is ready/merged and I can change this one.

**Comment by HannesWell** [comment]
> Please let me know when your PR is ready/merged and I can change this one.

Since https://github.com/eclipse-pde/eclipse.pde/pull/755 is now submitted, we can continue on this one. :)

**Comment by HannesWell** [comment]
> I suggest we add a method to `TracingOptionsManager` like `ensureInitialized()` that basically calls getTracingTemplate() and passes a `ProgressMonitor` to it so that we can at least see progress. It shouldn't be canclable because otherwise the state would become inconsistent or more complex to manage.

Thinking about it again, a `TracingOptionsManager.ensureInitialized(IProgressMonitor)` method could be canceled but then all results computed until then should be discarded.

To achieve this it is probably sufficient to replace `TracingOptionsManager.getTracingTemplate()` by 
```
	public void ensureInitialied(IProgressMonitor monitor) {
		getTracingTemplate(monitor);
	}

	private synchronized Map<String, String> getTracingTemplate(IProgressMonitor monitor) {
		if (template == null) {
			Map<String, String> temp = new HashMap<>();
			IPluginModelBase[] models = PluginRegistry.getAllModels();
			SubMonitor subMonitor = SubMonitor.convert(monitor, models.length);
			Arrays.stream(models).map(TracingOptionsManager::getOptions).filter(Objects::nonNull).forEach(p -> {
				subMonitor.split(1);
				@SuppressWarnings({ "rawtypes", "unchecked" })
				Map<String, String> entries = (Map) p;
				temp.putAll(entries); // All entries are of String/String
			});
			template = temp;
		}
		return template;
	}
```
and pass just null as monitor in all other callers of `getTracingTemplate()`.

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java
Comment (fedejeanne) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java]: I removed the cache but I left the internal class (it's now called `TracingOptionsManagerDelegate`). The reason for this is that the class provides a good encapsulation and helps me channel all calls to the "expensive methods" in `TracingOptionsManager`.

**Comment by fedejeanne** [comment]
I just pushed some changes and took your advice about using the `ProgressMonitor` inside the method `TracingOptionsManager::getTracingTemplate`, thank you for that! I made some adjustments though:
* I didn't add any method to make sure the template is initialized ( _e.g._ `ensureInitialized`) because I had no need for it
* I converted the stream into a `for`-loop to properly report the progress since one needs to call `monitor.split(1)` for every model and not for every _filtered_ (non-null) model. 
* I inverted the `if`-logic in the method to reduce the [cognitive complexity (nesting level)](https://docs.sonarsource.com/sonarqube/latest/user-guide/metric-definitions/).

## Important: do **NOT** merge yet
This PR is blocked by https://github.com/eclipse-platform/eclipse.platform/pull/766

**Comment by fedejeanne** [comment]
Test failure already documented here: https://github.com/eclipse-pde/eclipse.pde/issues/820

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java
Comment (HannesWell) [inline ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java:97]: Please move `subMonitor.split(1)` to the beginning of the loop.
`SubMonitor.split()` cleans up an previously created other children and therefore 'completes/calls `done()` for them.
Consequently calling split at the end of the loop has the effect that the first progress reporting for the loop happens after the second model has been processed because only then a pending submonitor-child exists.
This means a slight delay of the first feedbeck.

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java
Comment (HannesWell) [inline ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java:76]: If you prefer to have those methods bundled in a separate class (you could also just add a `IRunnableContext fRunnableContext` field to the `TracingBlock` class and have the methods there), I would make it a record and create a new record object whenever the context is reset (which I assume happens rarely).
```suggestion
	private static record TracingOptionsManagerDelegate(IRunnableContext runnableContext) {
```

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/PDEUIMessages.java
Comment (HannesWell) [inline ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/PDEUIMessages.java:1095]: This has no value defined and is never called.
```suggestion
```

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/PDEUIMessages.java
Comment (HannesWell) [inline ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/PDEUIMessages.java:1099]:
```suggestion
	public static String TracingBlock_initializing_tracing_options;
```
Because in all three of this cases the same method is called eventually for which the progress reporting happens I think it is sufficient to have one message here. The first one IMHO fits well. :)
```suggestion
	public static String TracingBlock_initializing_tracing_options;
```

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java
Comment (HannesWell) [inline ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java:107]: It would be nice if there would be some kind of `ICallableWithProgress`.

Comment (HannesWell) [inline ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java:107]: Instead of a `Function` you could use a `BiFunction<TracingOptionsManager, IProgressMonitor, Map<String, String>>` and move the repetitive calls of `PDECore.getDefault().getTracingOptionsManager()` to this method. This would even allow to use a Method-Reference in one caller.
Furthermore if always the same task-name is used, we would save passing that as an argument.
```suggestion
		private Map<String, String> runShowingProgress(
				BiFunction<TracingOptionsManager, IProgressMonitor, Map<String, String>> task) {
			try {
				final Map<String, String> result = new HashMap<>();
				// due to a bug
				// (https://github.com/eclipse-platform/eclipse.platform/issues/769),
				// the task needs to be forked otherwise the UI
				// does not show the progress indicator
				fRunnableContext.run(true, false, monitor -> {
					SubMonitor subMonitor = SubMonitor.convert(monitor,
							PDEUIMessages.TracingBlock_initializing_tracing_options, 1);
					result.putAll(task.apply(PDECore.getDefault().getTracingOptionsManager(), subMonitor.split(1)));
				});
```

**Review by HannesWell — COMMENTED** [review]
> I made some adjustments though:
> 
>     * I didn't add any method to make sure the template is initialized ( _e.g._ `ensureInitialized`) because I had no need for it
> 
> 
>     * I inverted the `if`-logic in the method to reduce the [cognitive complexity (nesting level)](https://docs.sonarsource.com/sonarqube/latest/user-guide/metric-definitions/).

That's fine.

>     * I converted the stream into a `for`-loop to properly report the progress since one needs to call `monitor.split(1)` for every model and not for every _filtered_ (non-null) model.

Good catch!




> I did **not** make the tasks/monitors cancelable for now because that requires to rethink how the UI behaves once the task has been canceled (probably some elements need to be disabled) and also where to handle the proper exceptions in the code. I'd like to leave that for a future PR since these changes already provide a good enough performance and usability improvement even without the task being cancelable

Sounds reasonable. If you one day make it cancalable, the Tracing-tab could look similar to the disabled tracing state, but the `Enable tracing` check-box state should not be affected by that. The editor could then contain a button instead to load the tracing options again.


With the mentioned changes this would be perfectly ready from PDE side.

> 
> This PR is blocked by [eclipse-platform/eclipse.platform#766](https://github.com/eclipse-platform/eclipse.platform/pull/766)

I'll try to participate there over the weekend.

**Comment by fedejeanne** [comment]
> Sounds reasonable. If you one day make it cancalable, the Tracing-tab could look similar to the disabled tracing state, but the `Enable tracing` check-box state should not be affected by that. The editor could then contain a button instead to load the tracing options again.

Thank you! I'll keep that in my for the next PR :-)
 
> I'll try to participate there over the weekend.

Thank you!

**Comment by fedejeanne** [comment]
Test failures documented in #554 and #844

File: ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java
Comment (fedejeanne) [reply ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TracingOptionsManager.java]: Done, thanks for the tip!

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/PDEUIMessages.java
Comment (fedejeanne) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/PDEUIMessages.java:1095]: Deleted

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/PDEUIMessages.java
Comment (fedejeanne) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/PDEUIMessages.java:1099]: Done

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java
Comment (fedejeanne) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java]: Done

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java
Comment (fedejeanne) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java]: I made it a record, thank you for the tip!

**Review by HannesWell — COMMENTED** [review]
Thanks for the adjustments.

I just have two non-critical remarks and then this is ready from the PDE site.

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java
Comment (fedejeanne) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java:74]: If it's ok to you, I'd rather keep it as it is: I find method references with parameters a bit harder to read so I usually go for the "verbose" option. The current style also matches the style of the other 2 calls in the class so that's another reason to keep it like this.

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java
Comment (fedejeanne) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java]: Good point, I changed it.

Better to have a descriptive exception than to let it throw a follow-up NPE. 

For the record: I will tackle the issue of showing the exception to the user in the UI when I make the task cancelable (in another PR) since then it will be much more likely that an exception happens (specifically: [an `OperationCanceledException` wrapped inside a `InterruptedException`](https://github.com/eclipse-platform/eclipse.platform.ui/pull/1259))

File: ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java
Comment (HannesWell) [reply ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/launcher/TracingBlock.java:74]: If you insist that's fine in this case :)

**Review by HannesWell — APPROVED** [review]
Great. So this looks ready from a PDE point of view.

Should we wait until your related Platform PR ist merged, before merging this?

**Comment by fedejeanne** [comment]
Yes please, let's wait until https://github.com/eclipse-platform/eclipse.platform/pull/766 is merged otherwise this PR will introduce a bug.

**Comment by fedejeanne** [comment]
@HannesWell https://github.com/eclipse-platform/eclipse.platform/pull/766 is merged --> this PR is good to go :)

**Comment by HannesWell** [comment]
Great!
Thanks for this and the platform fix.
Nice work :)

**Pull request merged**
